ocaml-lsp-server/src/dune.ml

```diff
@@ -233,43 +250,98 @@ end = struct
               let+ () = Progress.build_progress progress p in
               Some ()))
 
-  let diagnostic_loop client running dune_diagnostic_id diagnostics
-      ~include_promotions =
+  let registration_of_promotion (d : Drpc.Diagnostic.Promotion.t) =
+    let registrations =
+      let in_source = Drpc.Diagnostic.Promotion.in_source d in
+      let code_action =
+        let id = "ocamllsp-promote-action/" ^ in_source in
+        let method_ = "textDocument/codeAction" in
+        let registerOptions =
+          let documentSelector =
+            [ DocumentFilter.create ~pattern:in_source () ]
+          in
+          CodeActionRegistrationOptions.create ~documentSelector
+            ~codeActionKinds:[ CodeActionKind.Other "Promote" ]
+            ()
+          |> CodeActionRegistrationOptions.yojson_of_t
+        in
+        Registration.create ~id ~method_ ~registerOptions ()
+      in
+      [ code_action ]
+    in
+    RegistrationParams.create ~registrations
+
+  let fold_promotions promotions diagonstic ~f =
```

Comment:
What the relationship between promotions and a diagnostic?

Reply:
It's a parent-child relationship. Every diagnostic belongs to a parent. A single diagnostic may have multiple promotions.